README.md
# Genesis Header Nav

WordPress plugin that registers a menu location and displays it inside the header for a Genesis Framework child theme.

## Description

The default method of getting a menu to appear in the top right of a site using a child theme of the Genesis Framework, is to add a create a menu in Appearance -> Menus, and then place a Custom Menu widget in the Header Right widget area. While this works, it produces markup as seen in screenshot 1, and which has the following problems:

 * Since the menu is output from a widget, you end up with all of the extraneous widget and widget area markup - in a child theme with HTML5 support, that's the widget area `aside`, the widget `section`, and the widget wrap `div`. In themes without HTML5 support, it's three levels of `div` elements instead. Not only is this more DOM elements to render (performance), but all markup in the site header is pushing the real page content further down the source; search engines apparently put higher value on content at the top of the source (which is why Genesis ensures primary and secondary sidebars come lower in the source than the main content, irrespective of where they are displayed on screen).
 * In HTML5 themes, what could be a site's main navigation is wrapped in an `aside` element. It's not known whether this has any impact on SEO. Theoretically at least, search engines may put less value on navigation found in an `aside` or otherwise treat it differently.

    > "_I can't think of any good reason to use an aside in a header... what the hell would it be contextually related to? The logo? lol_" - **Robert Neu**

This plugin registers a new menu location called Header and, if a menu is assigned to it, displays it after the Header Right area. If you don't have any widgets in the Header Right area, then Genesis ensures that none of that widget area markup is output, so you end up with code like screenshot 2. If you do want a widget in the Header Right area, that's fine - it can be positioned and styled as you want, without negatively affecting the navigation menu as well.

## Screenshots

![Screenshot of markup using Custom Menu widget](assets/screenshot-1.png)  
_Screenshot 1: Markup using Custom Menu widget. Note the `aside`, `section` and `div` parents to `nav`._

---

![Screenshot of markup using this plugin](assets/screenshot-2.png)  
_Screenshot 2: Markup using this plugin. `nav` is a sibling element to the title area `div`._

## Requirements
 * PHP 5.3+
 * WordPress 3.9+
 * Genesis 2.1+

## Installation

### Upload

1. Download the latest tagged archive (choose the "zip" option).
2. Go to the __Plugins -> Add New__ screen and click the __Upload__ tab.
3. Upload the zipped archive directly.
4. Go to the Plugins screen and click __Activate__.

### Manual

1. Download the latest tagged archive (choose the "zip" option).
2. Unzip the archive.
3. Copy the folder to your `/wp-content/plugins/` directory.
4. Go to the Plugins screen and click __Activate__.

Check out the Codex for more information about [installing plugins manually](http://codex.wordpress.org/Managing_Plugins#Manual_Plugin_Installation).

### Git

Using git, browse to your `/wp-content/plugins/` directory and clone this repository:

`git clone git@github.com:GaryJones/genesis-header-nav.git`

Then go to your Plugins screen and click __Activate__.

## Updates

This plugin supports the [GitHub Updater](https://github.com/afragen/github-updater) plugin, so if you install that, this plugin becomes automatically updateable direct from GitHub.

## Change Log

See the [CHANGELOG.md](CHANGELOG.md).

## Usage

Once activated, head to Appearance -> Menus. Create a menu as usual, and assign it to the Header menu location.

## Backwards-incompatible Changes

The hook that filters the menu was called `genesis_do_header_nav` but is now called `genesis_header_nav` due to using the `genesis_header_nav()` function in Genesis 2.1.

**2.0.0:** Custom language files previously loaded from (example) `WP_LANG_DIR . '/genesis-header-nav/genesis-header-nav-en_GB.po'` now need to be placed at `WP_LANG_DIR . '/plugins/genesis-header-nav-en_GB.po'`, as per language packs.

**2.0.0:** This plugin uses PHP namespaces, so you'll need PHP 5.3+ powering your site.

## Customising

### CSS

The plugin should work with all Genesis child themes, though you may need to add styles to position the output in the traditional place of top right, e.g.:

~~~css
.nav-header {
	float: right;
	text-align: right;
	width: 50%;
}
~~~

Adjust the width as needed to allow enough space for your title area and menu items.

### Priority

The plugin includes a `genesis_header_nav_priority` filter, with a default value of `12`. Use the following values to reposition the nav accordingly:

* Before the `<header>` element, inside the `.site-container`: 0-4
* Before the title + widget area: 5-9
* After the title + widget area: 10-14
* After the `<header>` element: 15+

If you want to add it in between the title and widget area, you'll need to unhook and re-build `genesis_do_header()` function so that the output of the widget area is in a different function that can be hooked to a later priority.

As an example, to add the nav before the title + widget area markup in the source, you can use the following:

~~~php
add_filter( 'genesis_header_nav_priority', 'prefix_genesis_header_nav_priority' );
/**
 * Change the order of the nav within the header (Genesis Header Nav plugin).
 *
 * @param int $priority Existing priority. Default is 12.
 *
 * @return int New priority.
 */
function prefix_genesis_header_nav_priority( $priority ) {
	return 8;
}
~~~

### Top Menu (above `<header>`)

If you give the above priority filter a value of less than 5, then the output will be before the `<header>`, so that you can display what might be considered a Top menu. Of course, this might mean that the "Header" menu location label is confusing, but since that string is internationalised, it's possible to filter that and change it to make it easier for users to understand:

~~~php
add_filter( 'gettext', 'prefix_genesis_header_nav_name', 10, 3 );
/**
 * Change the name of the Header menu location added by Genesis Header Nav plugin.
 * 
 * @param string $translated_text Translated text.
 * @param string $original_text   Original text.
 * @param string $domain          Text domain.
 */
function prefix_genesis_header_nav_name( $translated_text, $original_text, $domain ) {
	if ( 'genesis-header-nav' === $domain && 'Header' === $original_text ) {
		return 'Top';
	}
	return $translated_text;
}
~~~

### Removing the Menu

If you want the menu to not display, perhaps on a landing page, then you can do the following:

~~~php
add_action( 'init', 'prefix_genesis_header_nav_remove', 11 );
/**
 * Remove Genesis Header Nav menu.
 */
function prefix_genesis_header_nav_remove() {
	if ( function_exists( 'Gamajo\GenesisHeaderNav\get_plugin' ) ) {
		remove_action( 'genesis_header', array( Gamajo\GenesisHeaderNav\get_plugin(), 'show_menu' ), apply_filters( 'genesis_header_nav_priority', 12 ) );
	}
}
~~~

## Credits

Built by [Gary Jones](https://twitter.com/GaryJ)  
Copyright 2013 [Gamajo Tech](http://gamajo.com/)


CHANGELOG.md
# Change Log for Genesis Header Nav

## [Unreleased]

_Nothing yet._

## [2.1.0] - 2016-08-10

* Add `composer.json`.
* Add this change log.
* Fix bug in documentation.

## [2.0.0] - 2014-11-19

_To be completed._

## [1.3.1] - 2014-11-08

* Fix for priority filter bug.

## [1.3.0] - 2014-10-30

_To be completed._

## [1.2.0] - 2013-09-07

_To be completed._

## [1.1.0] - 2013-08-30

* Add filter for priority of nav within header.

## 1.0.0 - 2013-08-30

* Initial release.

[Unreleased]: https://github.com/GaryJones/genesis-header-nav/compare/2.1.0...HEAD
[2.1.0]: https://github.com/GaryJones/genesis-header-nav/compare/2.0.0...2.1.0
[2.0.0]: https://github.com/GaryJones/genesis-header-nav/compare/1.3.1...2.0.0
[1.3.1]: https://github.com/GaryJones/genesis-header-nav/compare/1.3.0...1.3.1
[1.3.0]: https://github.com/GaryJones/genesis-header-nav/compare/1.2.0...1.3.0
[1.2.0]: https://github.com/GaryJones/genesis-header-nav/compare/1.1.0...1.2.0
[1.1.0]: https://github.com/GaryJones/genesis-header-nav/compare/1.0.0...1.1.0


genesis-header-nav.php
<?php
/**
 * Genesis Header Nav
 *
 * @package   Genesis_Header_Nav
 * @author    Gary Jones
 * @license   GPL-2.0+
 * @copyright 2013 Gary Jones, Gamajo Tech
 *
 * @wordpress-plugin
 * Plugin Name:       Genesis Header Nav
 * Plugin URI:        https://github.com/GaryJones/genesis-header-nav
 * Description:       Registers a menu location and displays it inside the header for a Genesis Framework child theme.
 * Version:           2.1.0
 * Author:            Gary Jones
 * Author URI:        https://gamajo.com
 * Text Domain:       genesis-header-nav
 * License:           GPL-2.0+
 * License URI:       http://www.gnu.org/licenses/gpl-2.0.txt
 * Domain Path:       /languages
 * GitHub Plugin URI: https://github.com/GaryJones/genesis-header-nav
 * GitHub Branch:     master
 * Requires WP:       3.9
 * Requires PHP:      5.3
 */

namespace Gamajo\GenesisHeaderNav;

// If this file is called directly, abort.
if ( ! defined( 'WPINC' ) ) {
	die;
}

// Main class file.
require_once plugin_dir_path( __FILE__ ) . 'includes/class-genesis-header-nav.php';

/**
 * Get plugin object.
 *
 * Instantiates it if necessary.
 *
 * @since 2.0.0
 *
 * @return Gamajo\GenesisHeaderNav\Plugin
 */
function get_plugin() {
	static $plugin = NULL;
	if ( is_null( $plugin ) ) {
		$plugin = new Plugin;
	}
	return $plugin;
}

add_action( 'plugins_loaded', __NAMESPACE__ . '\\genesis_header_nav_i18n' );
/**
 * Load Genesis Header Nav plugin text domain.
 *
 * @since 2.0.0
 */
function genesis_header_nav_i18n() {
	load_plugin_textdomain( 'genesis-header-nav', false, dirname( plugin_basename( __FILE__ ) ) . '/languages' );
}

add_action( 'after_setup_theme', __NAMESPACE__ . '\\genesis_header_nav_run' );
/**
 * Run Genesis Header Nav plugin.
 *
 * @since 2.0.0
 */
function genesis_header_nav_run() {
	$plugin = get_plugin();
	$plugin->run();
}


includes/class-genesis-header-nav.php
<?php
/**
 * Genesis Header Nav
 *
 * @package   Genesis_Header_Nav
 * @author    Gary Jones
 * @license   GPL-2.0+
 * @copyright 2013 Gary Jones, Gamajo Tech
 */

namespace Gamajo\GenesisHeaderNav;

/**
 * Plugin class.
 *
 * @package Genesis_Header_Nav
 * @author  Gary Jones
 */
class Plugin {
	/**
	 * Initialize the plugin by setting localization, filters, and administration functions.
	 *
	 * Applies `genesis_header_nav_priority` filter. Use a value of 6-9 to add the nav before title + widget area, or
	 * 11-14 to add it after. If you want to add it in between, you'll need to remove and re-build `genesis_do_header()`
	 * so that the output of the widget area is in a different function that can be hooked to a later priority.
	 *
	 * @since 1.0.0
	 */
	public function run() {
		add_action( 'init', array( $this, 'register_nav_menu' ) );
		add_action( 'genesis_header', array( $this, 'show_menu' ), apply_filters( 'genesis_header_nav_priority', 12 ) );
		add_filter( 'body_class', array( $this, 'body_classes' ), 15 );
	}

	/**
	 * Register the menu location.
	 *
	 * @since 1.0.0
	 */
	public function register_nav_menu() {
		register_nav_menu( 'header', __( 'Header', 'genesis-header-nav' ) );
	}

	/**
	 * Display the menu.
	 *
	 * @since 1.0.0
	 */
	public function show_menu() {
		$class = 'menu genesis-nav-menu menu-header';
		if ( genesis_superfish_enabled() ) {
			$class .= ' js-superfish';
		}

		genesis_nav_menu(
			array(
				'theme_location' => 'header',
				'menu_class'     => $class,
			)
		);
	}

	/**
	 * Remove then conditionally re-add header-full-width body class.
	 *
	 * As well as just checking for something being in the header right widget area, or the action having something
	 * hooked* in, we also need to check to see if the header navigation has a menu assigned to it. Only if all are
	 * false can we* proceed with saying the header-full-width class should be applied.
	 *
	 * Function must be hooked after priority 10, so Genesis has had a chance to do the filtering first.
	 *
	 * @since 1.0.0
	 *
	 * @param array $classes Existing classes.
	 *
	 * @return array Amended classes.
	 */
	public function body_classes( array $classes ) {
		// Loop through existing classes to remove 'header-full-width'
		foreach ( $classes as $index => $class ) {
			if ( 'header-full-width' === $class ) {
				unset( $classes[$index] );
				break; // No need to check the rest.
			}
		}

		// Do all the checks
		if ( ! is_active_sidebar( 'header-right' ) && ! has_action( 'genesis_header_right' ) && ! has_nav_menu( 'header' ) )
			$classes[] = 'header-full-width';

		return $classes;
	}
}
